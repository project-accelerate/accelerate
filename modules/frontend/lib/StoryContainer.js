import React from 'react'
import Theme from '../components/Theme'

export function StoryContaner({ children }) {
  return (
    <div>
      <Theme />
      {children}
    </div>
  )
}
