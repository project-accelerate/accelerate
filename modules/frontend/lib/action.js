import * as React from "react";
import { wrapDisplayName } from "recompose";
import { mapValues } from "lodash";

/**
 * Utility HOC for handling async actions.
 *
 * Given a map of async functions for handling events, the events will be
 * passed into the wrapped component as props.
 *
 * The event handler may return an object containing result data.
 *
 * Each event handler will have:
 *
 *  * A property `pending`, indecating whether the action is in progress
 *  * A property, `error` if the action failed
 *  * Any properties in the result object.
 *
 * @param {object} actionPropMap
 */
export function actions(actionPropMap) {
  return WrappedComponent =>
    class Actions extends React.PureComponent {
      static displayName = wrapDisplayName(WrappedComponent, "actions");

      state = { actionState: {} };

      get actions() {
        return mapValues(actionPropMap, (handler, actionName) => {
          const action = (...args) => {
            this.perform(actionName, handler, args);
          };

          return Object.assign(action, this.state.actionState[actionName]);
        });
      }

      async perform(actionName, handler, args) {
        try {
          this.actionWillStart(actionName);
          const result = await handler(this.props, ...args);

          this.actionDidComplete(actionName, result);
        } catch (error) {
          this.actionDidFail(actionName, error);
        }
      }

      actionWillStart(action) {
        const { actionState } = this.state;
        this.setState({
          actionState: { ...actionState, [action]: { pending: true } }
        });
      }

      actionDidComplete(action, result) {
        const { actionState } = this.state;
        this.setState({
          actionState: {
            ...actionState,
            [action]: { pending: false, ...result }
          }
        });
      }

      actionDidFail(action, error) {
        const { actionState } = this.state;
        this.setState({
          actionState: { ...actionState, [action]: { pending: false, error } }
        });
      }

      render() {
        return <WrappedComponent {...this.props} {...this.actions} />;
      }
    };
}
