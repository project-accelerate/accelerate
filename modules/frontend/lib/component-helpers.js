import React from 'react'
import PropTypes from 'prop-types'
import classnames from 'classnames'

/**
 * Utility for defining a container component (one that has no layout logic and simply
 * applies a css class to a container.
 */
export function createWrapper(name, { type, baseClass }) {
  function CardComponent({ children, className }) {
    return React.createElement(
      type,
      { className: classnames(baseClass, className) },
      ...(Array.isArray(children) ? children : [children])
    )
  }

  CardComponent.displayName = name
  
  CardComponent.propTypes = {
    className: PropTypes.string,
    children: PropTypes.string,
  }

  CardComponent.defaultProps = {
    className: undefined,
    children: undefined
  }

  return CardComponent
}

