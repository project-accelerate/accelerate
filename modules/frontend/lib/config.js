/**
 * Get development config from appropriate source.
 *
 * When running on server, we read from process.ENV.
 *
 * When running in browser, we read from window.ENV, which contains
 * whitelisted configuration variables passed through from the server.
 *
 * We use this approach rather than baking config vars into the js bundle
 * to support Heroku pipelines, which serve the same js bundles regardless
 * of which stage in the pipeline we're at.
 */

module.exports = process.browser ? window.ENV : process.env;
