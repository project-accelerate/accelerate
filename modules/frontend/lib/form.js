import * as React from "react";
import { wrapDisplayName, hoistStatics } from "recompose";
import { fromPairs, pickBy } from "lodash";

/**
 * Utility HOC for binding form fields.
 *
 * This adds the following props to the wrapped component:
 *
 *   * a `fields` property containing form binding props suitable for
 *      attaching to an input element
 *
 *   * a `formValue` property containing form value as a single object.
 *
 *   * a `formIsValid` property indicating whether all fields pass validation.
 *
 *   * a `fieldErrors`property  mapping field ids to validation errors.
 *
 * @param {object} fields Array of ids for each field
 *
 * @param {function} validate
 *  Validator function. Should return either:
 *
 *    * Boolean indicating success or failure
 *  or:
 *    * Object mapping field keys to validation errors
 *
 */
export function form({ fields, validate = () => {} }) {
  return hoistStatics(
    WrappedComponent =>
      class Form extends React.Component {
        static displayName = wrapDisplayName(WrappedComponent, "form");

        state = {
          values: fromPairs(fields.map(key => [key, ""])),
          blurred: {}
        };

        get validationErrors() {
          const validationResult = validate(this.state.values) || {};
          return pickBy(validationResult, (_, key) => this.state.blurred[key]);
        }

        get isValid() {
          const validationErrors = validate(this.state.values);
          return (
            validationErrors && !fields.some(field => validationErrors[field])
          );
        }

        get fields() {
          return fromPairs(
            fields.map(field => [
              field,
              {
                onChange: event => {
                  this.setState({
                    values: {
                      ...this.state.values,
                      [field]: event.currentTarget.value
                    }
                  });
                },
                onBlur: () => {
                  this.setState({
                    blurred: {
                      ...this.state.blurred,
                      [field]: true
                    }
                  });
                },
                value: this.state.values[field]
              }
            ])
          );
        }

        render() {
          return (
            <WrappedComponent
              {...this.props}
              fields={this.fields}
              formValue={this.state.values}
              formIsValid={this.isValid}
              fieldErrors={this.validationErrors}
            />
          );
        }
      }
  );
}
