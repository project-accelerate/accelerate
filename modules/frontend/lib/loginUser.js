import fetch from "isomorphic-fetch";

export async function loginUser({ username, password }) {
  try {
    const result = await fetch("/login", {
      method: "POST",
      body: JSON.stringify({ username, password }),
      headers: {
        "Content-Type": "application/json"
      },
      credentials: "same-origin"
    });

    if (result.status === 401) {
      return { succeeded: false, reason: E_INVALID_CREDENTIALS };
    }

    if (!result.ok) {
      return { succeeded: false, reason: E_UNEXPECTED_ERROR };
    }

    return { succeeded: true };
  } catch (error) {
    return { succeeded: false, reason: E_NETWORK_ERROR };
  }
}

export const E_INVALID_CREDENTIALS = "invalid credentials";
export const E_UNEXPECTED_ERROR = "unexpected error";
export const E_NETWORK_ERROR = "network error";
