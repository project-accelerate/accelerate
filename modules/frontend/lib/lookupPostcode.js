import fetch from "isomorphic-fetch";

export async function lookupPostcode({ longitude, latitude }) {
  const result = await fetch(
    `http://api.postcodes.io/postcodes?lon=${longitude}&lat=${latitude}`
  );
  if (!result.ok) {
    return null;
  }

  const { result: hits } = await result.json();
  return hits && Array.isArray(hits) && hits[0].postcode;
}

export async function lookupPostcodeFromUserLocation() {
  const coords = await getUserLocation();
  if (coords) {
    return lookupPostcode(coords);
  }

  return null;
}

function getUserLocation() {
  return new Promise((resolve, reject) => {
    if (!navigator.geolocation) {
      return resolve();
    }

    return navigator.geolocation.getCurrentPosition(
      ({ coords }) => resolve(coords),
      reject,
      { timeout: 10000 }
    );
  });
}
