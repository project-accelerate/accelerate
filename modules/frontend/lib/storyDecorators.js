import React, { Component } from "react";
import PropTypes from "prop-types";

/**
 * Wraps a storybook story in a relatively positioned div of a given
 * height so that it can be sized by its parent without necessarily
 * taking up the whole screen in storybook.
 */
export function withHeight(height) {
  return storyFn => (
    <div style={{ position: "relative", height }}>{storyFn()}</div>
  );
}

/**
 * Storybook data wrapper to bind the onChange and value props
 * of a component so that state changes are shown in the storybook.
 */
export function withStateWrapper(storyFn) {
  return <StateWrapper>{storyFn()}</StateWrapper>;
}

class StateWrapper extends Component {
  static propTypes = {
    children: PropTypes.element.isRequired,
    defaultValue: PropTypes.any.isRequired
  };

  state = { value: this.props.defaultValue };

  handleChange = change => {
    this.setState({ value: change });
  };

  render() {
    return React.cloneElement(
      React.Children.only(this.props.children),
      { value: this.state.value, onChange: this.handleChange },
      null
    );
  }
}
