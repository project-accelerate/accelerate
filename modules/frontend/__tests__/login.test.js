/* eslint-env jest */

import React from "react";
import { mount } from "enzyme";
import renderer from "react-test-renderer";
import fetch from "isomorphic-fetch";
import Router from "next/router";
import waitUntil from "async-wait-until";
import LoginPage from "../pages/login";
import { E_INVALID_CREDENTIALS } from "../lib/loginUser";

describe("/login", () => {
  const CORRECT_EMAIL = "me@example.com";
  const CORRECT_PASSWORD = "1234";

  beforeEach(() => {
    fetch.mockImplementation(async (url, req) => {
      const { username, password } = JSON.parse(req.body);
      if (username === CORRECT_EMAIL && password === CORRECT_PASSWORD) {
        return new Response(undefined, { status: 201 });
      }

      return new Response(undefined, { status: 401 });
    });
  });

  it("should match snapshot", () => {
    const tree = renderer.create(<LoginPage />);
    expect(tree.toJSON()).toMatchSnapshot();
  });

  it("should redirect user in when username and password are accepted", async () => {
    await logInWithCredentials({
      username: CORRECT_EMAIL,
      password: CORRECT_PASSWORD
    });

    expect(Router.replace).toHaveBeenCalledWith("/");
  });

  it("should remain on page and show error on failure", async () => {
    const tree = await logInWithCredentials({
      username: CORRECT_EMAIL,
      password: "invalid password"
    });

    expect(Router.replace).not.toHaveBeenCalled();
    expect(tree.containsMatchingElement(E_INVALID_CREDENTIALS)).toBeTruthy();
  });

  async function logInWithCredentials({ username, password }) {
    const tree = mount(<LoginPage />);
    const emailField = tree.find("input#email");
    const passwordField = tree.find("input#password");
    const login = tree.find("button#submit");

    emailField
      .getElement()
      .props.onChange({ currentTarget: { value: username } });
    passwordField
      .getElement()
      .props.onChange({ currentTarget: { value: password } });
    login.simulate("click");

    await waitUntil(() => {
      tree.update();
      const page = tree.find("LoginPage").getElement();
      return !page.props.onSubmit.pending;
    });

    return tree;
  }
});
