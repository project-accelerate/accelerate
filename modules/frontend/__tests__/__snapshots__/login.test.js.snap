// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`/login should match snapshot 1`] = `
<form
  method="none"
  onSubmit={[Function]}
>
  <div
    aria-describedby={undefined}
    className="MuiFormControl-root-1 MuiFormControl-marginNormal-2 MuiFormControl-fullWidth-4"
    onBlur={[Function]}
    onFocus={[Function]}
    variant="email"
  >
    <label
      className="MuiFormLabel-root-11 MuiInputLabel-root-5 MuiInputLabel-formControl-6 MuiInputLabel-animated-9"
      data-shrink={false}
      htmlFor="email"
    >
      Email
    </label>
    <div
      className="MuiInput-root-15 MuiInput-fullWidth-23 MuiInput-formControl-16 MuiInput-inkbar-17 MuiInput-underline-21"
      onBlur={[Function]}
      onFocus={[Function]}
    >
      <input
        aria-invalid={false}
        aria-required={false}
        autoComplete="email"
        autoFocus={undefined}
        className="MuiInput-input-24"
        defaultValue={undefined}
        disabled={false}
        id="email"
        name={undefined}
        onChange={[Function]}
        onKeyDown={undefined}
        onKeyUp={undefined}
        placeholder={undefined}
        readOnly={undefined}
        required={undefined}
        rows={undefined}
        type="text"
        value=""
      />
    </div>
  </div>
  <div
    aria-describedby={undefined}
    className="MuiFormControl-root-1 MuiFormControl-marginNormal-2 MuiFormControl-fullWidth-4"
    onBlur={[Function]}
    onFocus={[Function]}
    variant="password"
  >
    <label
      className="MuiFormLabel-root-11 MuiInputLabel-root-5 MuiInputLabel-formControl-6 MuiInputLabel-animated-9"
      data-shrink={false}
      htmlFor="password"
    >
      Password
    </label>
    <div
      className="MuiInput-root-15 MuiInput-fullWidth-23 MuiInput-formControl-16 MuiInput-inkbar-17 MuiInput-underline-21"
      onBlur={[Function]}
      onFocus={[Function]}
    >
      <input
        aria-invalid={false}
        aria-required={false}
        autoComplete="current-password"
        autoFocus={undefined}
        className="MuiInput-input-24 MuiInput-inputType-27"
        defaultValue={undefined}
        disabled={false}
        id="password"
        name={undefined}
        onChange={[Function]}
        onKeyDown={undefined}
        onKeyUp={undefined}
        placeholder={undefined}
        readOnly={undefined}
        required={undefined}
        rows={undefined}
        type="password"
        value=""
      />
    </div>
  </div>
  <button
    className="MuiButtonBase-root-45 MuiButtonBase-disabled-46 MuiButton-root-30 MuiButton-disabled-39"
    disabled={true}
    id="submit"
    onBlur={[Function]}
    onClick={[Function]}
    onFocus={[Function]}
    onKeyDown={[Function]}
    onKeyUp={[Function]}
    onMouseDown={[Function]}
    onMouseLeave={[Function]}
    onMouseUp={[Function]}
    onTouchEnd={[Function]}
    onTouchMove={[Function]}
    onTouchStart={[Function]}
    tabIndex={-1}
    type="button"
  >
    <span
      className="MuiButton-label-31"
    >
      Log In
    </span>
  </button>
</form>
`;
